### PR TITLE
Issue282 handling syntax errors in projects config

### LINKED ISSUE #282: Handling errors in projects.cfg
Some syntax errors of  `projects.cfg` produce a lengthy generic Python error message, see e.g. https://groups.google.com/d/msg/annif-users/kIvJPbZ-vVU/zXDYFm6PAgAJ 

A more concise/specific error message would be better. Some errors are already handled, for example this exists: 
`"Error: Project 'maui-fi': vocab setting is missing"`

### PR DESCRIPTION
**Pull request by juhoinkinen**
Raise `ConfigurationException` instead of full traceback in following situations:

1. when trying to use (train/suggest) a project without a backend set in `projects.cfg`:
    ```
    $ annif train nobackend -p tests/projects.cfg any_trainingdata.tsv 
    Error: Project 'nobackend': backend setting is missing
    ```
    This is in line with missing vocab error: ` "Error: Project 'maui-fi': vocab setting is missing"`.

2. when trying to use `projects.cfg` that has duplicated sections or entries:
    ```
    $ annif train any_backend -p tests/projects_invalid.cfg any_trainingdata.tsv 
    Error: Error: While reading from 'tests/projects_invalid.cfg' [line 10]: section 'duplicatedproject' already exists
    ```
This closes #282.

### CODE REVIEW

**Comment by osma** [comment]
Travis Python 3.5 build is failing, I think VW needs to be upgraded first. Maybe merge #296?

**Comment by osma** [comment]
failing build log: https://travis-ci.org/NatLibFi/Annif/jobs/559957266

**Comment by osma** [comment]
Looks good, but there is one further case that could be covered. If `analyzer` is left unspecified, I get this:

```
  File "/home/oisuomin/git/Annif/annif/project.py", line 188, in _create_vectorizer
    tokenizer=self.analyzer.tokenize_words)
AttributeError: 'NoneType' object has no attribute 'tokenize_words'
```

Should this case be covered too, or left for another issue/PR?

**Comment by juhoinkinen** [comment]
> Looks good, but there is one further case that could be covered. If `analyzer` is left unspecified, I get this:
> 
> ```
>   File "/home/oisuomin/git/Annif/annif/project.py", line 188, in _create_vectorizer
>     tokenizer=self.analyzer.tokenize_words)
> AttributeError: 'NoneType' object has no attribute 'tokenize_words'
> ```
> 
> Should this case be covered too, or left for another issue/PR?

I'll check if this can be easily covered too and implement, otherwise I'll make a new issue about this.

**Comment by juhoinkinen** [comment]
> Looks good, but there is one further case that could be covered. If `analyzer` is left unspecified, I get this:
> 
> ```
>   File "/home/oisuomin/git/Annif/annif/project.py", line 188, in _create_vectorizer
>     tokenizer=self.analyzer.tokenize_words)
> AttributeError: 'NoneType' object has no attribute 'tokenize_words'
> ```
> 
> Should this case be covered too, or left for another issue/PR?

Implemented raising `ConfigurationError` for this in the previous commit. For the test I changed the [`[noanalyzer]` project](https://github.com/NatLibFi/Annif/pull/300/commits/f03c320435fb1d460e99bd36144a4ac78a43e734#diff-e889023e3a656e930515ec0cb95ce5b1R36) to use TFIDF backend because dummy backend does not need analyzer and would not raise the `ConfigurationError`. That project was included for issue #148. I did not see any reason for the backend to remain `dummy`, but could @Osma confirm?

File: annif/project.py [hunk 1]
```diff
@@ -120,8 +123,14 @@ def _suggest_with_backend(self, text, backend_params):
 
     @property
     def analyzer(self):
-        if self._analyzer is None and self.analyzer_spec:
-            self._analyzer = annif.analyzer.get_analyzer(self.analyzer_spec)
+        if self._analyzer is None:
+            if self.analyzer_spec:
+                self._analyzer = annif.analyzer.get_analyzer(
+                    self.analyzer_spec)
+            elif self.backend.needs_subject_vectorizer:
```
Comment (osma) [inline annif/project.py]: I don't think this is the right condition to test, maybe not even the right place for it. It happens to work for the case where an analyzer is needed within the AnnifProject class, but there are other situations in backend-specific code where an analyzer is needed as well, irrespective of `needs_subject_vectorizer`:

* [fasttext.py](https://github.com/NatLibFi/Annif/blob/master/annif/backend/fasttext.py#L77)
* [mixins.py](https://github.com/NatLibFi/Annif/blob/master/annif/backend/mixins.py#L21)
* [vw_multi.py](https://github.com/NatLibFi/Annif/blob/master/annif/backend/vw_multi.py#L57)

Alternatively, backends could have a `needs_analyzer` field that indicates that they are making use of the analyzer, then the `if` clause above could check for that.

File: annif/project.py [hunk 2]
```diff
@@ -62,10 +62,13 @@ def _init_access(self):
...
-        analyzer = self.analyzer
-        logger.debug("Project '%s': initialized analyzer: %s",
-                     self.project_id,
-                     str(analyzer))
+        try:
+            analyzer = self.analyzer
+            logger.debug("Project '%s': initialized analyzer: %s",
+                         self.project_id,
+                         str(analyzer))
+        except AnnifException as err:
```
Comment (osma) [inline annif/project.py]: Can we avoid having to catch the exception here? Not sure it's a great idea that simply accessing `self.analyzer` may trigger an exception. Returning None is a bit more elegant, but then of course the missing analyzer case has to be handled by the calling code.

**Pull request merged**